ktor-utils/jvm/src/io/ktor/util/pipeline/ExceptionConstructor.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+package io.ktor.util.pipeline
+
+import kotlinx.coroutines.*
+import java.lang.reflect.*
+import java.util.*
+import java.util.concurrent.locks.*
+import kotlin.concurrent.*
+
+private val throwableFields = Throwable::class.java.fieldsCountOrDefault(-1)
+private val cacheLock = ReentrantReadWriteLock()
+private typealias Ctor = (Throwable) -> Throwable?
+
+// Replace it with ClassValue when Java 6 support is over
+private val exceptionCtors: WeakHashMap<Class<out Throwable>, Ctor> = WeakHashMap()
+
+@Suppress("UNCHECKED_CAST")
+internal fun <E : Throwable> tryCopyException(exception: E, cause: Throwable): E? {
```

Comment:
Please comment that it's a complete copy-paste of `kx-coroutines` API to override the cause